lib/active_fedora/orders/ordered_list.rb

```diff
@@ -0,0 +1,237 @@
+      def empty?
+        head.next == tail
+      end
+
+      # @param [ActiveFedora::Base] target Target to append to list.
+      # @option [::RDF::URI, ActiveFedora::Base] :proxy_in Proxy in to 
+      #   assert on the created node.
+      def append_target(target, proxy_in: nil)
+        node = build_node(new_node_subject)
+        node.target = target
+        node.proxy_in = proxy_in
+        append_to(node, tail.prev)
+      end
+
+      # @param [Integer] loc Location to insert target at
+      # @param [ActiveFedora::Base] target Target to insert
+      def insert_at(loc, target)
+        node = build_node(new_node_subject)
+        node.target = target
+        if loc == 0
+          append_to(node, head)
+        else
+          append_to(node, ordered_reader.take(loc).last)
+        end
+      end
+
+      # @param [ListNode] node Node to delete
+      def delete_node(node)
+        node = ordered_reader.find{|x| x == node}
+        if node
+          prev_node = node.prev
+          next_node = node.next
+          node.prev.next = next_node
+          node.next.prev = prev_node
+          @changed = true
+        end
+        self
+      end
+
+      # @param [Integer] loc Index of node to delete.
+      def delete_at(loc)
+        return self if loc == nil
+        arr = ordered_reader.take(loc+1)
+        if arr.length == loc+1
+          delete_node(arr.last)
+        else
+          self
+        end
+      end
+
+      # @return [Boolean] Whether this list was changed since instantiation.
+      def changed?
+        @changed
+      end
+
+      # @return [::RDF::Graph] Graph representation of this list.
+      def to_graph
+        ::RDF::Graph.new.tap do |g|
+          array = to_a
+          array.map(&:to_graph).each do |resource_graph|
+            g << resource_graph
+          end
+        end
+      end
+
+      # Marks this list as not changed.
+      def changes_committed!
+        @changed = false
+      end
+
+      private
+
+      attr_reader :node_cache
+
+      def append_to(source, append_node)
+        source.prev = append_node
+        if append_node.next
+          append_node.next.prev = source
+          source.next = append_node.next
+        else
+          self.tail = source
+        end
+        append_node.next = source
+        @changed = true
+      end
+
+      def ordered_reader
+        ActiveFedora::Aggregation::OrderedReader.new(self)
+      end
+
+      def build_node(subject=nil)
+        return nil unless subject
+        node_cache.fetch(subject) do
+          ActiveFedora::Orders::ListNode.new(node_cache, subject, graph)
+        end
+      end
+
+      def new_node_subject
+        node = ::RDF::URI("##{::RDF::Node.new.id}")
+        while(node_cache.has_key?(node))
+          node = ::RDF::URI("##{::RDF::Node.new.id}")
+        end
+        node
+      end
+
+      class NodeCache
+        def initialize
+          @cache ||= {}
+        end
+
+        def fetch(uri)
```

Comment:
So if the cache doesn't contain the given URI and no block is provided, this method should return nil?